include/revng/Model/VerifyHelper.h

```diff
@@ -41,6 +42,11 @@ class VerifyHelper {
 
   ~VerifyHelper() { revng_assert(InProgress.size() == 0); }
 
+public:
+  bool hasPushedTracking() const { return HasPushedTracking; }
+
+  void setPushedTracking() { HasPushedTracking = true; }
```

Comment:
2) i fail to see how is it fundamentally different now 

```cpp
bool Binary::verifyTypes(VerifyHelper &VH) const {
  revng::TrackingPushingScope Dc(VH, *this);
```

i did it this way because if i return a lock from a function it is implementation defined if it pushes once or it pushes, pops, and pushes again, due to rvo.

3) i considered it, but it can't be done, the fundamentally try to solve different problems, one is trying shutdown the tracking of a whole structure and is intended to let you push multiple times, assuming that it will not happen too many times. The other is instead used to prevent mutually recursive heterogeneous functions from running out of available pushes when they don't care about tracking, which are just 7.

Reply:
2. I think implementing the move constructor can solve the RVO problem you're mentioning.
3. `VH.disableTracking` should first check if it has already been disabled. If it has been disabled, it should return a no-op `TrackGuard` (as I suggested, we can have a `bool` in the constructor to make `TrackGuard` a no-op). Otherwise, it returns a regular `TrackGuard`.
   Basically I'm suggesting to move part of the implementation of the `TrackingPushingScope` constructor, into `VH.disableTracking`. I don't see why it can't be done.